app/src/org/commcare/dalvik/activities/CommCareSetupActivity.java

```diff
@@ -579,32 +584,33 @@ private void performSMSInstall(boolean installTriggeredManually){
     }
 
 
-
     /**
-     * Scan the SMS inbox, looking for messages that meet the expected install format,
-     * and if found and verified return the discovered install link. Current behavior will search
-     * backwards from the most recent text, returning the first discovered valid link
-     * @return the verified install link, null if none found
-     * @throws SignatureException if we discovered a valid-looking message but could not verifyMessageSignatureHelper it
+     * Scan the most recent incoming text messages for a message with a
+     * verified link to a commcare app and install it.  Message scanning stops
+     * after the number of scanned messages reaches 'SMS_CHECK_COUNT'.
+     *
+     * @param installTriggeredManually don't install the found app link
      */
     private void scanSMSLinks(boolean installTriggeredManually){
         // http://stackoverflow.com/questions/11301046/search-sms-inbox
         final Uri SMS_INBOX = Uri.parse("content://sms/inbox");
         Cursor cursor = getContentResolver().query(SMS_INBOX, null, null, null, "date desc");
+        if (cursor == null) {
+            return;
+        }
+        int messageIterationCount = 0;
         try {
-            if (cursor.moveToFirst()) { // must check the result to prevent exception
-                while (!cursor.isAfterLast()) {
-                    String textMessageBody = cursor.getString(cursor.getColumnIndex("body"));
-                    if (textMessageBody.contains(GlobalConstants.SMS_INSTALL_KEY_STRING)) {
-                        RetrieveParseVerifyMessageTask mTask =
-                                new RetrieveParseVerifyMessageTask(this,installTriggeredManually);
-                        mTask.execute(textMessageBody);
-                        break;
-                    }
+            while (cursor.moveToNext() && messageIterationCount <= SMS_CHECK_COUNT) {
+                String textMessageBody = cursor.getString(cursor.getColumnIndex("body"));
+                messageIterationCount++;
+                if (textMessageBody.contains(GlobalConstants.SMS_INSTALL_KEY_STRING)) {
+                    RetrieveParseVerifyMessageTask mTask =
+                            new RetrieveParseVerifyMessageTask(this, installTriggeredManually);
+                    mTask.execute(textMessageBody);

```

Comment:
@wpride 

This should probably be wrapped in 

        if (Build.VERSION.SDK_INT >= Build.VERSION_CODES.HONEYCOMB) {
            mProcess.executeOnExecutor(AsyncTask.THREAD_POOL_EXECUTOR, records);
        } else {
            mProcess.execute(records);
        }


by the way, also shouldn't this task be connected to this activity using the task framework.